File: HackerRank/Hard/MaximumSubarraySum.cs
using System;
using System.Collections.Generic;

namespace HackerRank.Hard
{
    //https://www.hackerrank.com/challenges/maximum-subarray-sum/problem

    public class MaximumSubarraySum
    {
        private static long _max;

        public static long CalcMax(long[] a, long devider)
        {
            for (var i = 0; i < a.Length; i++)
            {
                long sum = a[i];

                CheckMax(devider, sum);

                for (var j = i; j < a.Length; j++)
                {
                    sum += a[j];
                    CheckMax(devider, sum);
                }
                if (_max == devider - 1)
                    return _max;
            }

            return _max;
        }

        private static void CheckMax(long devider, long sum)
        {
            var tmpTes = sum % devider;

            if (tmpTes > _max)
                _max = tmpTes;
        }

        public static long GetMax() => _max;
    }
}

File: HackerRankCpp/MaximumSubarraySum.cpp
#include "stdafx.h"
#include "MaximumSubarraySum.h"
#include <iostream>
#include <set>
#include <algorithm>
using namespace std;

//https://www.hackerrank.com/challenges/maximum-subarray-sum/problem

MaximumSubarraySum::MaximumSubarraySum()
{
}

long long MaximumSubarraySum::Solve(long long n, long long m, long long * a)
{
	long long x, prefix = 0, maxim = 0;
	set<long long> S;
	S.insert(0);
	for (int i = 1; i <= n; i++) {
		x = a[i];
		prefix = (prefix + x) % m;
		maxim = max(maxim, prefix);
		set<long long>::iterator  it = S.lower_bound(prefix + 1);
		if (it != S.end()) {
			maxim = max(maxim, prefix - (*it) + m);
		}
		S.insert(prefix);
	}
	return maxim;
}



MaximumSubarraySum::~MaximumSubarraySum()
{
}
